com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.refactor.cs

```diff
@@ -0,0 +1,262 @@
+                // Upgrade prefab instance copies of referencables only
+                var conversionLinks = new List<ConversionLink>();
+                var timelineManager = new TimelineManager(scene);
+                var allPrefabInstances = new List<GameObject>();
+
+                for (var p = 0; p < m_PrefabManager.PrefabCount; ++p)
+                {
+                    allPrefabInstances.AddRange(
+                        PrefabManager.FindAllInstancesOfPrefabEvenInNestedPrefabs(scene,
+                            m_PrefabManager.GetPrefabAssetPath(p)));
+                }
+
+                var upgradedObjects = new HashSet<GameObject>();
+                var upgradable =
+                    GetUpgradables(allPrefabInstances.ToArray(), m_ObjectUpgrader.RootUpgradeComponentTypes, true);
+                foreach (var go in upgradable)
+                {
+                    if (upgradedObjects.Contains(go))
+                        continue; // Ignore if already converted (this can happen in nested prefabs)
+                    upgradedObjects.Add(go);
+#if CINEMACHINE_TIMELINE
+                    var originalVcam = go.GetComponent<CinemachineVirtualCameraBase>();
+                    var timelineReferences = timelineManager.GetTimelineReferences(originalVcam);
+#endif
+                    var convertedCopy = Object.Instantiate(go);
+                    UpgradeObjectComponents(convertedCopy, null);
+                    m_ObjectUpgrader.DeleteObsoleteComponents(convertedCopy);
+
+                    var conversionLink = new ConversionLink
+                    {
+                        originalName = go.name,
+                        originalGUIDName = GUID.Generate().ToString(),
+                        convertedGUIDName = GUID.Generate().ToString(),
+#if CINEMACHINE_TIMELINE
+                        timelineReferences = timelineReferences,
+#endif
+                    };
+                    go.name = conversionLink.originalGUIDName;
+                    PrefabUtility.RecordPrefabInstancePropertyModifications(go); // record name change
+
+                    convertedCopy.name = conversionLink.convertedGUIDName;
+                    conversionLinks.Add(conversionLink);
+                }
+
+                conversionLinksPerScene.Add(s, conversionLinks);
+
+                EditorSceneManager.SaveScene(scene);
+            }
+
+            m_CurrentSceneOrPrefab = string.Empty;
+        }
+        
+        /// <summary>
+        /// For each prefab asset that has any component from filter
+        /// - Upgrade the prefab asset, but do not delete obsolete components
+        /// - Fix timeline references
+        /// - Fix object references
+        /// - Fix animation references
+        /// </summary>
+        /// <param name="upgradeReferencables">
+        /// True, then only referencable prefab instances are upgraded.
+        /// False, then only non-referencable prefab instances are upgraded.</param>
```

Comment:
What if a prefab has both a referencable and a non-referencable?  e.g. dolly vcam + path in the same prefab.  I think you have to revisit the logic in this function to handle those cases correctly.

Reply:
Because when all is upgraded. I do a pass to correct references before obsoletes are deleted.

Reply:
hmmm, I was wondering what that was for (cough cough hack cough)